File: payload_crypto.py
"""AES-256-GCM encryption for app payload assets.

Phase A of docs/SECURITY_CDR_PIPELINE.md: the Pi encrypts CDR-bearing assets
before uploading them to the public GitHub Release, so the release exposes
ciphertext only. Disabled unless AR_LOCAL_PAYLOAD_ENC=1 (the app must ship
decrypt support — Phase B — before this is switched on).

Requires the `cryptography` package, which Raspberry Pi OS ships system-wide
(Debian: python3-cryptography). Imported lazily so unencrypted builds keep the
repo's stdlib-only footprint.

Asset format (``.json.gz.enc``)::

    ARE1 | 12-byte nonce | AES-256-GCM ciphertext+tag (AAD = b"ARE1")

The nonce is derived as HMAC-SHA256(key, sha256(plaintext))[:12]. Identical
(key, plaintext) pairs therefore produce identical bytes — required because
same-day rebuilds must stay content-addressable (the app skips re-download on
unchanged sha256). A nonce can only repeat for an identical plaintext, which
yields the identical ciphertext, so GCM nonce-reuse leakage cannot occur.
"""
from __future__ import annotations

import hashlib
import hmac
import os
from pathlib import Path
from typing import Optional

MAGIC = b"ARE1"
NONCE_LEN = 12
KEY_LEN = 32
ALG = "aes-256-gcm"

ENV_FLAG = "AR_LOCAL_PAYLOAD_ENC"
ENV_KEY_FILE = "AR_LOCAL_PAYLOAD_KEY_FILE"
DEFAULT_KEY_FILE = "/etc/ar-local/payload.key"


def _aesgcm_cls():
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    except ImportError as exc:  # pragma: no cover - environment-specific
        raise RuntimeError(
            "payload encryption requires the 'cryptography' package "
            "(Debian/Raspberry Pi OS: apt install python3-cryptography)"
        ) from exc
    return AESGCM


def load_key(path: Path) -> bytes:
    """Read a 32-byte key stored as 64 hex chars (whitespace tolerated)."""
    key = bytes.fromhex(path.read_text(encoding="utf-8").strip())
    if len(key) != KEY_LEN:
        raise ValueError(
            f"payload key must be {KEY_LEN} bytes (64 hex chars), got {len(key)} bytes"
        )
    return key


def key_id(key: bytes) -> str:
    """Short non-secret identifier so clients can confirm they hold the right key."""
    return hashlib.sha256(b"ar-local-payload-key:" + key).hexdigest()[:8]


def resolve_key_from_env() -> Optional[bytes]:
    """Key when AR_LOCAL_PAYLOAD_ENC is truthy, else None (encryption off)."""
    if (os.environ.get(ENV_FLAG) or "").strip().lower() not in {"1", "true", "yes"}:
        return None
    return load_key(Path(os.environ.get(ENV_KEY_FILE) or DEFAULT_KEY_FILE))


def _derive_nonce(key: bytes, plain: bytes) -> bytes:
    return hmac.new(key, b"nonce:" + hashlib.sha256(plain).digest(), hashlib.sha256).digest()[
        :NONCE_LEN
    ]


def encrypt_asset(plain: bytes, key: bytes) -> bytes:
    nonce = _derive_nonce(key, plain)
    return MAGIC + nonce + _aesgcm_cls()(key).encrypt(nonce, plain, MAGIC)


def decrypt_asset(blob: bytes, key: bytes) -> bytes:
    if blob[: len(MAGIC)] != MAGIC:
        raise ValueError("not an ARE1 encrypted asset")
    nonce = blob[len(MAGIC) : len(MAGIC) + NONCE_LEN]
    return _aesgcm_cls()(key).decrypt(nonce, blob[len(MAGIC) + NONCE_LEN :], MAGIC)

File: mobile/src/lib/payloadCrypto.ts
import { gcm } from '@noble/ciphers/aes';
import { hexToBytes } from '@noble/ciphers/utils';
import { sha256 } from '@noble/hashes/sha256';

import { PAYLOAD_DEC_KEY_HEX } from '../config';

/**
 * Decrypt support for AES-256-GCM payload assets produced by the Pi's
 * payload_crypto.py (Phase B of docs/SECURITY_CDR_PIPELINE.md).
 *
 * Asset format: `ARE1 | 12-byte nonce | GCM ciphertext+tag` with AAD = "ARE1".
 * The key currently ships in app config (interim — obfuscation, not security);
 * Phase D replaces it with auth-gated key issuance.
 */

const MAGIC = Uint8Array.from([0x41, 0x52, 0x45, 0x31]); // "ARE1"
const NONCE_LEN = 12;
const KEY_LEN = 32;

export function isEncryptedAsset(bytes: Uint8Array): boolean {
  return (
    bytes.length > MAGIC.length + NONCE_LEN &&
    bytes[0] === MAGIC[0] &&
    bytes[1] === MAGIC[1] &&
    bytes[2] === MAGIC[2] &&
    bytes[3] === MAGIC[3]
  );
}

/** Short non-secret key identifier; must match manifest `enc.key_id`. */
export function payloadKeyId(keyHex: string): string {
  const prefix = 'ar-local-payload-key:'; // ASCII; avoids a TextEncoder dependency
  const key = hexToBytes(keyHex);
  const buf = new Uint8Array(prefix.length + key.length);
  for (let i = 0; i < prefix.length; i += 1) buf[i] = prefix.charCodeAt(i);
  buf.set(key, prefix.length);
  return Array.from(sha256(buf).slice(0, 4))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

export function decryptAsset(bytes: Uint8Array, keyHex: string = PAYLOAD_DEC_KEY_HEX): Uint8Array {
  if (!keyHex) {
    throw new Error('payload asset is encrypted but no decryption key is configured');
  }
  const key = hexToBytes(keyHex);
  if (key.length !== KEY_LEN) {
    throw new Error(`payload decryption key must be ${KEY_LEN} bytes (64 hex chars)`);
  }
  if (!isEncryptedAsset(bytes)) {
    throw new Error('not an ARE1 encrypted asset');
  }
  const nonce = bytes.slice(MAGIC.length, MAGIC.length + NONCE_LEN);
  const ciphertext = bytes.slice(MAGIC.length + NONCE_LEN);
  return gcm(key, nonce, MAGIC).decrypt(ciphertext);
}
